devdoc/cancellation_token_requirements.md

```diff
@@ -0,0 +1,138 @@
+}
+
+// create a cancellation token
+THANDLE(CANCELLATION_TOKEN) token = cancellation_token_create(false);
+ASSERT_IS_NOT_NULL(token);
+
+// check that it hasn't been canceled
+bool canceled;
+int result = cancellation_token_is_canceled(token, &canceled)
+ASSERT_ARE_EQUAL(int, 0, result);
+ASSERT_IS_FALSE(canceled);
+
+// register to be notified when the token is canceled
+THANDLE(CANCELLATION_TOKEN_REGISTRATION) registration = cancellation_token_register_notify(token, on_cancel, NULL);
+ASSERT_IS_NOT_NULL(registration);
+
+// signal cancellation - causes 'on_cancel' to be called
+result = cancellation_token_cancel(token);
+ASSERT_ARE_EQUAL(int, 0, result);
+
+// check that token has been canceled
+result = cancellation_token_is_canceled(token, &canceled)
+ASSERT_ARE_EQUAL(int, 0, result);
+ASSERT_IS_TRUE(canceled);
+
+// clean up
+THANDLE_ASSIGN(CANCELLATION_TOKEN_REGISTRATION)(&registration, NULL);
+THANDLE_ASSIGN(CANCELLATION_TOKEN)(&token, NULL);
+```
+
+## Exposed API
+
+```c
+typedef struct CANCELLATION_TOKEN_TAG CANCELLATION_TOKEN;
+THANDLE_TYPE_DECLARE(CANCELLATION_TOKEN);
+
+typedef struct CANCELLATION_TOKEN_REGISTRATION_TAG CANCELLATION_TOKEN_REGISTRATION;
+THANDLE_TYPE_DECLARE(CANCELLATION_TOKEN_REGISTRATION);
+
+MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN), cancellation_token_create, bool, canceled);
+MOCKABLE_FUNCTION(, int, cancellation_token_is_canceled, THANDLE(CANCELLATION_TOKEN), cancellation_token, bool*, canceled);
+MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN_REGISTRATION), cancellation_token_register_notify, THANDLE(CANCELLATION_TOKEN), cancellation_token, TCALL_DISPATCHER_TARGET_FUNC_TYPE_NAME(CANCELLATION_TOKEN_CANCEL_CALL), on_cancel, void*, context);
+MOCKABLE_FUNCTION(, int, cancellation_token_cancel, THANDLE(CANCELLATION_TOKEN), cancellation_token);
+```
+
+## cancellation_token_create
+
+```c
+MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN), cancellation_token_create, bool, canceled);
+```
+
+`cancellation_token_create` creates a new cancellation token with the specified initial cancellation state.
+
+**SRS_CANCELLATION_TOKEN_04_001: [** `cancellation_token_create` shall allocate memory for a `THANDLE(CANCELLATION_TOKEN)`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_002: [** If any underlying error occurs `cancellation_token_create` shall fail and return `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_003: [** `cancellation_token_create` shall create a `TCALL_DISPATCHER` handle by calling `TCALL_DISPATCHER_CREATE(CANCELLATION_TOKEN_CANCEL_CALL)`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_004: [** `cancellation_token_create` shall set the initial state to be equal to `canceled` parameter. **]**
+
+**SRS_CANCELLATION_TOKEN_04_005: [** `cancellation_token_create` shall return a valid `THANDLE(CANCELLATION_TOKEN)` when successful. **]**
+
+## cancellation_token_is_canceled
+
+```c
+MOCKABLE_FUNCTION(, int, cancellation_token_is_canceled, THANDLE(CANCELLATION_TOKEN), cancellation_token, bool*, canceled);
```

Comment:
>int [](http://example.com/codeflow?start=20&length=3)

I'm tempted to suggest this just returns bool. Simplifies the calling code slightly since this really can't fail unless the token is NULL. If a NULL is given, returning false makes sense (a NULL token is not canceled).

Reply:
/consider of course

Reply:
Done.